### PR TITLE
Store shared token only if newer than what's in config

### PR DESCRIPTION
**Pull request by jlorper**
### Fixes
- Tries to minimize race conditions when storing the shared token.
NOTE: Test pending

### Change implications

 - dependencies added/changed? **no**

### CODE REVIEW

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java [hunk 1]
```diff
@@ -205,9 +210,11 @@ public AccessToken refreshAccessToken() throws IOException {
             }
 
             if (shouldRefresh(this.currentToken, clock.instant())) {
+                // some jitter to minimize multiple instances refreshing at same time
+                Uninterruptibles.sleepUninterruptibly(50 + RandomUtils.nextLong(50, 200), TimeUnit.MILLISECONDS);
```
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:214]: there's already a `RandomNumberGenerator` being injected above - L182

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java [hunk 2]
```diff
@@ -348,17 +355,41 @@ Optional<AccessToken> getSharedAccessTokenIfSupported() {
...
-                        objectMapper.writerFor(AccessTokenDto.class)
-                            .writeValueAsString(AccessTokenDto.toAccessTokenDto(accessToken)));
-                    log.log(Level.INFO, "New token stored in config");
-                } catch (JsonProcessingException e) {
-                    log.log(Level.SEVERE, "Could not serialize token into JSON", e);
+                // at least synchronize in the same JVM
+                synchronized ($refreshLock) {
+                    try {
+                        // wait slightly before checking for most recent token in config
+                        Uninterruptibles.sleepUninterruptibly(RandomUtils.nextLong(50, 200), TimeUnit.MILLISECONDS);
```
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:365]: again, already a `RandomNumberGenerator` being injected above; use that, can fake it to 1 for tests or something

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java [hunk 3]
```diff
@@ -205,9 +210,11 @@ public AccessToken refreshAccessToken() throws IOException {
             }
 
             if (shouldRefresh(this.currentToken, clock.instant())) {
+                // some jitter to minimize multiple instances refreshing at same time
+                Uninterruptibles.sleepUninterruptibly(50 + RandomUtils.nextLong(50, 200), TimeUnit.MILLISECONDS);
                 tokenResponse = exchangeRefreshTokenForAccessToken();
-                this.currentToken = asAccessToken(tokenResponse);
-                storeSharedAccessTokenIfSupported(this.currentToken);
+                AccessToken newToken = asAccessToken(tokenResponse);
+                this.currentToken = storeSharedAccessTokenIfSupported(newToken);
```
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:217]: more readable to just have `this.currentToken = newToken`, then `storedSharedAccessTokenIfSupported(newToken)` as separate line

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java [hunk 4]
```diff
@@ -205,9 +210,11 @@ public AccessToken refreshAccessToken() throws IOException {
             }
```
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:210]: possibly everything above here should be moved, as it's not actually refreshing the token.

Comment (eschultink) [reply java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:210]: and actually, this is `refreshAccessTokenIfNeeded` or something - as it won't *necessarily* refresh.

**Review by eschultink — COMMENTED** [review]
nowhere quick to do it, but I think `currentToken` on this class is in effect acting to "cache" the token until it expires. in shared case, we might want to be periodically reloading the token from parameter store in case updated - regardless of TTL of token itself

**Comment by jlorper** [comment]
Gonna discard, after a thorough review the jitter in the refresh token threshold is more than enough. We might do the lock to minimize writes on initial burst, but pretty negligible. (no 401 errors seen at all, besides initial ones)

**Pull request closed without merging**